### DISCUSSION TITLE
How to detach a session from a connection

### DISCUSSION THREAD

**Discussion by QasimK** [sqlalchemy/sqlalchemy#11192] [Usage Questions]
I would like to detach a session from a connection in a similar way to how objects are detached from a session, so that I can implement the following pseudo code.

```
session.attach(connection)

try:
    # my read queries here
finally:
    session.detach()
```

I am using SQLAlchemy 2, the connection is from a connection pool, and it is in `AUTOCOMMIT` mode. When the connection is returned to the pool (on `detach`) it may be reused by other sessions implementing the above. I'm connecting to Postgresql and no transaction will occur.

I want the session itself to retain its identity map so that perhaps it could be continued to be used later.

I have considered `expunge_all` but that would be a decent amount of work and/or overhead. However, this could be one route to implementation if needed.

I wondering if there are any (internal/private) methods that would help or if anyone has to implement this before. I'm willing to hack Session and deal with breaking changes.

**Comment by CaselIT** [comment]
Hi,

What you want to do kinda sounds like creating a session from a connectiom with `expire_on_commit=False` then terminating the work with commit.
That would leave the connection not at the pool, but you could manually close it.

Something like
```py
session = Session(connection, expire_on_commit=False)
...
session.commit()
connection.close()
```
